syntax="proto2";

import "raft/enum.proto";

package raft;
option cc_generic_services = true;


// data store in baidu-rpc's attachment
message EntryMeta {
    required int64 term = 1;
    required EntryType type = 2;
    repeated string peers = 3;
    optional int64 data_len = 4;
};

message RequestVoteRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required int64 last_log_term = 5;
    required int64 last_log_index = 6;
};

message RequestVoteResponse {
    required int64 term = 1;
    required bool granted = 2;
};

message AppendEntriesRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required int64 prev_log_term = 5;
    required int64 prev_log_index = 6;
    repeated EntryMeta entries = 7;
    required int64 committed_index = 8;
};

message AppendEntriesResponse {
    required int64 term = 1;
    required bool success = 2;
    optional int64 last_log_index = 3;
};

message InstallSnapshotRequest {
    required string group_id = 1;
    required string server_id = 2;
    required string peer_id = 3;
    required int64 term = 4;
    required int64 last_included_log_term = 5;
    required int64 last_included_log_index = 6;
    repeated string peers = 7;
    required string uri = 8;
};

message InstallSnapshotResponse {
    required int64 term = 1;
    required bool success = 2;
};


service RaftService {
    rpc pre_vote(RequestVoteRequest) returns (RequestVoteResponse);

    rpc request_vote(RequestVoteRequest) returns (RequestVoteResponse);

    rpc append_entries(AppendEntriesRequest) returns (AppendEntriesResponse);

    rpc install_snapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
};
